backend/layers/processing/dataset_metadata_update.py

```diff
@@ -0,0 +1,233 @@
+import json
+import logging
+import os
+from typing import Dict
+
+import scanpy
+import tiledb
+from rpy2.robjects import StrVector
+from rpy2.robjects.packages import importr
+
+from backend.common.utils.corpora_constants import CorporaConstants
+from backend.common.utils.tiledb import consolidation_buffer_size
+from backend.layers.business.business import BusinessLogic
+from backend.layers.common.entities import (
+    CollectionVersionId,
+    DatasetArtifactType,
+    DatasetConversionStatus,
+    DatasetProcessingStatus,
+    DatasetStatusKey,
+    DatasetValidationStatus,
+    DatasetVersion,
+    DatasetVersionId,
+)
+from backend.layers.persistence.persistence import DatabaseProvider
+from backend.layers.processing.logger import configure_logging
+from backend.layers.processing.process_download import ProcessDownload
+from backend.layers.thirdparty.s3_provider import S3Provider
+from backend.layers.thirdparty.uri_provider import UriProvider
+
+configure_logging(level=logging.INFO)
+
+base = importr("base")
+seurat = importr("SeuratObject")
+
+
+class DatasetMetadataUpdate(ProcessDownload):
+    def __init__(
+        self, business_logic: BusinessLogic, artifact_bucket: str, cellxgene_bucket: str, datasets_bucket: str
+    ) -> None:
+        super().__init__(business_logic, business_logic.uri_provider, business_logic.s3_provider)
+        self.artifact_bucket = artifact_bucket
+        self.cellxgene_bucket = cellxgene_bucket
+        self.datasets_bucket = datasets_bucket
+
+    def update_h5ad(
+        self,
+        h5ad_uri: str,
+        original_dataset_version: DatasetVersion,
+        key_prefix: str,
+        new_dataset_version_id: DatasetVersionId,
+        metadata_update_dict: Dict[str, str],
+    ):
+        h5ad_filename = self.download_from_source_uri(
+            source_uri=h5ad_uri,
+            local_path=CorporaConstants.LABELED_H5AD_ARTIFACT_FILENAME,
+        )
+
+        adata = scanpy.read_h5ad(h5ad_filename)
+        metadata = original_dataset_version.metadata
+        for key, val in metadata_update_dict.items():
+            adata.uns[key] = val
+            if hasattr(metadata, key):
+                setattr(metadata, key, val)
+        adata.write(h5ad_filename, compression="gzip")
+        self.business_logic.set_dataset_metadata(new_dataset_version_id, metadata)
+
+        self.create_artifact(
+            h5ad_filename,
+            DatasetArtifactType.H5AD,
+            key_prefix,
+            new_dataset_version_id,
+            self.artifact_bucket,
+            DatasetStatusKey.H5AD,
+            datasets_bucket=self.datasets_bucket,
+        )
+        self.update_processing_status(
+            new_dataset_version_id, DatasetStatusKey.VALIDATION, DatasetValidationStatus.VALID
+        )
+        self.update_processing_status(new_dataset_version_id, DatasetStatusKey.H5AD, DatasetConversionStatus.CONVERTED)
+
+    def update_rds(
+        self,
+        rds_uri: str,
+        key_prefix: str,
```

Comment:
I think the `new` vs `original` approach you've taken for some of the var names in this PR is a good one. I'd say apply it broadly: I'd rename this to `new_key_prefix`.

Also consider `old` instead of `original`. Though tbh I'm not sure I care too much between those two.

Reply:
old/new, previous/current